### ISSUE TITLE
Access to arbitrary tables

### ISSUE THREAD

**Issue by steffilazerte** [BirdsCanada/naturecounts#9]
Eventually we want to be able to give BS employees (at least) access to arbitrary tables through the R package.

This will require changes to the API as well as the package.

**Comment by denislepage** [comment]
We already did some work towards this, and we should try to close the loop. Some of it may already be supported in the API with this entry point.

/api/query

E.g.

/api/query?table=bmde_filter_bad_dates&project_id=1005

The following tables are available now:

api_tables: list of tables (or views) that can be accessed through the API
                table_name: resource name exposed to the API
                akn_level: user level required for access
                local_name: local table/view name if different from table_name. I think this may also support dbname.dbo.table_name format to access data outside dendroica, but we’ll need to verify.

api_tables_filters: list of field names for api_tables that can be used to filter results (e.g. project_id).

api_user_roles: list of user permissions to access those data tables. We have 2 roles, but I don’t quite remember the distinction: READ and READTABLE. I believe READ is related to global access to BMDE data (e.g. Danielle could access all BMDE data, including level 0), and READTABLE is specifically for the custom tables. I’m not sure the current API makes that distinction now.

Right now, tables are assigned a level, and anyone at that level or lower can access those resources. E.g. admins (level 0) and can access all tables. bmde_filter_bad_date is set at level 3. Steffi, your user should be able to see it (you are level 3 for READTABLE). Resources set at level 5 should be accessible by anyone without special permissions as long as they have a token. There are none currently, but we can add one for testing.

We may need to add an API call that would return the available resources and query parameters that can be used based on the user access. There’s only 1 available so far.

I don’t think we’ll need to get more granular with permissions for now but that may be something we’ll need later (e.g. assigning individual tables to people in a more targeted way, so we can give someone access to a single level 0 table, for instance).

**Comment by steffilazerte** [comment]
(Adapted from the email chain)

Now the syntax for API Entry points is:

- https://sandbox.birdscanada.org/api/query/api_tables
    - The api_tables one will return the list of tables for the token’s user, the supported query parameters, and the required parameters.
- https://sandbox.birdscanada.org/api/query/bmde_filter_bad_dates?token=1234
    - e.g., https://sandbox.birdscanada.org/api/query/bmde_filter_bad_dates?token=asdfasdf&filter={"project_id":1005,"SiteCode":"DMBO"}
- https://sandbox.birdscanada.org/api/query/test (Requires parameter for language: lang=EN or FR)

 https://sandbox.birdscanada.org/api/query Should return the same as /query/api_tables. 

We want a simple function where the user can supply a table name, and query arguments. 

@pmorrill @denislepage I do get the api tables returned when I use /query/api_tables, but I get an error using just /query (also supplying a token):

"org.bsc.exceptions.DBException com.microsoft.sqlserver.jdbc.SQLServerException: Procedure or function 'prAPI_list_tables' expects parameter '@user_id', which was not supplied. SQL: \"EXEC dendroica.dbo.prAPI_list_tables\"

**Comment by steffilazerte** [comment]
@denislepage I've got the first pass at the function `nc_query_table()` on sandbox (https://github.com/BirdStudiesCanada/naturecounts/blob/sandbox/R/query_table.R). I can return the table data, no problem for all tables.

@pmorrill For some reason the filters I supply to `bmde_filter_bad_dates` always result in returning zero data. On the other hand I have no problem passing `lang = "EN"` for the `test` table. 

As far as I can see the filter format is the same as for the download api so it *should* work. Is there something I'm missing? I can give you more details if that would help.

**Comment by steffilazerte** [comment]
@denislepage, do you have any more information on what I'm doing wrong here?

**Comment by denislepage** [comment]
Do you want to try again? @steffilazerte.

This is now working as expected I believe: 
https://sandbox.birdscanada.org/api/query/bmde_filter_bad_dates?filter={%22project_id%22:1005,%22SiteCode%22:%22DMBO%22}

I also fixed a problem with the api_tables request. If no authentication is provided, you only get the public tables (2 at the moment). https://sandbox.birdscanada.org/api/query/api_tables

**Comment by steffilazerte** [comment]
Great, that works!

Now I'm having trouble supplying more than one option to a specific filter. 

For example, one species id works:
7680: https://sandbox.birdscanada.org/api/query/bmde_filter_bad_dates?filter={%22species_id%22:7680}
9750: https://sandbox.birdscanada.org/api/query/bmde_filter_bad_dates?filter={%22species_id%22:9750}

But not two species ids:
7680 and 9750: https://sandbox.birdscanada.org/api/query/bmde_filter_bad_dates?filter={%22species_id%22:[%227680%22,%229750%22]}

Is this because the JSON format expected is 'unboxed'? I.e. unboxed: `{"species_id":7680}` as opposed to boxed: `{"species_id":[7680]}` which means I can't give two options?

**Comment by steffilazerte** [comment]
By the way, are there any private tables I can test with ("testuser")? Right now I only get public tables with no username, "sample", "steffilazerte" and "testuser".

**Comment by denislepage** [comment]
I needed to dig up a bit, but it doesn’t look like we have set up table-specific permission levels at this point. We may eventually need to have this option, but for now, we have access levels for tables and users.

The steffilazerte user should have permission to see the table called “test”, which is a level 3 (same as your user).

If you are not seeing it, then there’s an issue that we’ll need to address.

**Comment by steffilazerte** [comment]
Hmm, no I'm not seeing it, when use 'steffilazerte' and look at the list of tables (either by querying table =  "api_tables"  or table = NULL, I only get "bmde_filter_bad_dates" as a table (other than "api_tables").

**Comment by steffilazerte** [comment]
Okay, current status (adapted from email). I've split this into two sections: listing tables and pulling data from tables. 
@denislepage could you take a look at the **Questions/Problems/Additions** for each section and let me know what you think?

---

**List Tables**

- Unauthenticated: https://sandbox.birdscanada.org/api/query?table=api_tables
- Authenticated: https://sandbox.birdscanada.org/api/query?token=token&trace=1&table=api_tables

`naturecounts` equivalent:

```
> nc_query_table()
             table_name                        filters       required
1 bmde_filter_bad_dates project_id,SiteCode,species_id       NA

> nc_query_table(username = "steffilazerte")
             table_name                        filters       required
1 bmde_filter_bad_dates project_id,SiteCode,species_id       NA
```
> **Questions/Problems/Additions**
> - Shouldn't the `test` table be listed when pulling the authenticated list of tables?
> - What is `trace = 1`? Do we need it? Doesn't seem to do anything on my end...

---

**Return data from tables**

- Unauthenticated all data: https://sandbox.birdscanada.org/api/query?trace=1&table=bmde_filter_bad_dates
- Unauthenticated filtered data: https://sandbox.birdscanada.org/api/query?trace=1&table=bmde_filter_bad_dates&filter={"project_id":1013}
- Authenticated missing required parameter: https://sandbox.birdscanada.org/api/query?token=token&trace=1&table=test
- Authenticated with required parameter: https://sandbox.birdscanada.org/api/query?token=token&trace=1&table=test&filter={"lang":"fr"}

**More filters**


- **Works** using species: https://sandbox.birdscanada.org/api/query?token=token&trace=1&table=test&filter={"species_id":9750} 
- **Doesn't work** using mult species: https://sandbox.birdscanada.org/api/query?token=token&trace=1&table=test&filter={"species_id":[7680,9750]} 
   - No error, but returns no data. Perhaps using AND rather than OR?

`naturecounts` equivalent:
```
# unauthenticated - no problem
> d <- nc_query_table(table = "bmde_filter_bad_dates")

# authenticated - error, as expected
> d <- nc_query_table(table = "test")
 Error: NatureCounts API request returned an error 
Message: 'Table does not allow querying through api'
API: https://sandbox.birdscanada.org/api/query

# unauthenticated - species filter
> d <- nc_query_table(table = "bmde_filter_bad_dates",  species_id = 7680, username = "sample")
```
>  **Questions/Problems/Additions**
> - What is `trace = 1` for? It doesn't seem to be necessary...
> - Perhaps trying to access a table the user doesn't have access to could give a more informative error?
> - Multiple options to a single filter should use OR for filtering (i.e. should be able to give multiple species ids)

**Comment by denislepage** [comment]
`    Shouldn't the test table be listed when pulling the authenticated list of tables?`

Yes, that would be the expected response. Will investigate.

`    What is trace = 1? Do we need it? Doesn't seem to do anything on my end...`

In theory, this adds more details in the response. Looking at the code, I don't know if this is implemented in all cases. I know this works for the mobile API.  Based on the code I am looking at, it only adds logging on the server side, but doesn't return anything in the request. You can probably keep it there for now, as it may help me debug issues.

Note that there is also a parameter called "debug=1". It seems to be returning the SQL script generated by the query, instead of returning data, but the intent is to remove this once the feature is working. 

`Multiple options to a single filter should use OR for filtering (i.e. should be able to give multiple species ids)`

Generally, multiple values for the same parameter will be a IN statement (A or B or C), and the boolean between different parameters should be an AND.  For instance: SpeciesID in (1,2,3) AND year = 2010.  I think the issue here is more likely related to handling of the array format, which is currently not supported (I think).

`Perhaps trying to access a table the user doesn't have access to could give a more informative error?`

Perhaps so. The response from the API is likely going to be a 403 (authentication issue), but I can check if we are also returning a message that the R package can display.

**Comment by denislepage** [comment]
Quick update: the system is currently designed to tied a table, a view or a proc to the entry points. Tables and views could be made to support filter arrays, but they are not at the moment. Procs however do not support arrays as input parameters. I think that in order to support arrays, we'd have to provide extra information at the field level to indicate for which fields this is supported. A workaround for now is for the user to make distinct calls for each species, or get all the data and filter locally. I'm inclined to leave things as is for now, unless @DMEthier identifies a critical need for this right away.

**Comment by denislepage** [comment]
OK, the permission issue for api_tables appears resolved in sandbox, and authenticated calls for authorized users should now return the "test" table. @steffilazerte can you test?

Also, I potentially have found a way to accept array parameters in procs if needed. The tables and views would be easier to generalize, so that any parameter could in theory be sent as a single value or an array, but the procs could not, and we'd have to either have a way to identify which parameters support arrays, and which do not, or perhaps have a distinct parameter that is always expected to be an array (e.g. speciesID and speciesIDArray as 2 distinct filters). The latter would be simpler for the API end, as this is how the database queries will need to operate. .

I'm still inclined not to enable this for now in favour of having the user (mostly staff for now) do their own filtering locally, but it's useful to know at least that the option exists, and a good reminder for me that views would be preferable over procs when possible in case we want to implement this later.

**Comment by steffilazerte** [comment]
Thanks for checking on this @denislepage!

1. Yes, the `test` table is listed now, great!
2. I'll keep the `trace` argument then
3. Would it be worth adding the multiple filter options to the R side of things? It's pretty straightforward to query the API multiple times if any given filter has more than one value? That way it's an option, but keeps the API side of things a bit simpler?

Essentially just this under the hood:
```
> d1 <- nc_query_table(table = "bmde_filter_bad_dates",
+                      species_id = 15770, username = "sample")
> d2 <- nc_query_table(table = "bmde_filter_bad_dates",
+                      species_id = 9750, username = "sample")
> rbind(d1, d2)
  survey_month species_id period                               comments project_id SiteCode survey_year start_year
1            7      15770      0 sp removed due to inadequate coverage        1005      IBS        2014          0
2            8      15770      0 sp removed due to inadequate coverage        1005      IBS        2014          0
3            9      15770      0 sp removed due to inadequate coverage        1005      IBS        2014          0
4           10      15770      0 sp removed due to inadequate coverage        1005      IBS        2014          0
5            7       9750      0 sp removed due to inadequate coverage        1005      IBS        2014          0
6            8       9750      0 sp removed due to inadequate coverage        1005      IBS        2014          0
7            9       9750      0 sp removed due to inadequate coverage        1005      IBS        2014          0
8           10       9750      0 sp removed due to inadequate coverage        1005      IBS        2014          0
  end_year survey_day
1        0          0
2        0          0
3        0          0
4        0          0
5        0          0
6        0          0
7        0          0
8        0          0
```

**Comment by denislepage** [comment]
I think we're good to deploy the new package then?

**Comment by denislepage** [comment]
For some reason. my earlier response by email didn't make it here:

You could handle this with multiple calls, but it could get significant if you combine multiple arrays (e.g. 5 years, 10 projects and 100 species = 5000 API calls).

It’s probably easier to get all data (depends on the table, but so far, most of the ones we will expose through this are of relatively small size), and then apply a local filter after the fact, which could also be done in the background on the R side, of course.

It wouldn’t be highly difficult to implement something more flexible on the server-side, if we think this is a tool that people will need.

**Comment by DMEthier** [comment]
RE: "I'm inclined to leave things as is for now, unless @DMEthier identifies a critical need for this right away."

I can pull the full data table (e.g., bad dates) and locally filter as needed in R. Thus, a multi-species array is not critical on my end.

**Comment by denislepage** [comment]
You could handle this with multiple calls, but it could get significant if you combine multiple arrays (e.g. 5 years, 10 projects and 100 species = 5000 API calls).

It’s probably easier to get all data (depends on the table, but so far, most of the ones we will expose through this are of relatively small size), and then apply a local filter after the fact, which could also be done in the background on the R side, of course.

It wouldn’t be highly difficult to implement something more flexible on the server-side, if we think this is a tool that people will need.

**Comment by steffilazerte** [comment]
Closed in [v0.3.0](https://github.com/BirdsCanada/naturecounts/tree/v0.3.0)
